app/src/main/java/com/eveningoutpost/dexdrip/utils/LibreTrendUtil.java

```diff
@@ -82,71 +115,71 @@ boolean IsTimeValid(long timeId) {
         return timeId >= 0 && timeId < MAX_POINTS;
     }
     
-    List<LibreTrendPoint> getData(long startTimestamp, long endTimestamp) {
-        Log.i(TAG, "Size of array is " + m_points.size() + " this = " + this);
+    public List<LibreTrendPoint> getData(long startTimestamp, long endTimestamp) {
+        Log.i(TAG, "Size of array is " + m_points.size() + " this = " + this + "m_libreTrendLatest.timestamp " + m_libreTrendLatest.timestamp);
         
         long startTime = Math.max(startTimestamp, m_libreTrendLatest.timestamp);
-        List<LibreBlock> latestBlocks = LibreBlock.getForTrend(startTime, endTimestamp);
+        // The extra 1 is to make sure we don't read the last packet again and again.
+        List<LibreBlock> latestBlocks = LibreBlock.getForTrend(startTime+1, endTimestamp);
         
         Log.i(TAG, "Size of latestBlocks is " + latestBlocks.size());
+        if(latestBlocks.size() > 0) {
+            Log.i(TAG, "Last packet timestamp is " + latestBlocks.get(latestBlocks.size()-1).timestamp);
+        }
         
         // Go for the last libreBlock and get calculated bg and timestamp.
-        if (latestBlocks.size() > 0) {
-            LibreBlock lastBlock = latestBlocks.get(latestBlocks.size() - 1);
-            // 
-            ReadingData readingData = NFCReaderX.getTrend(lastBlock);
-            if(readingData == null){
-                Log.e(TAG, "Error: NFCReaderX.getTrend retuned null for latest block");
-                return m_points;
-            }
-            if(readingData.trend.size() > 0 ) {
-                m_libreTrendLatest.id = (int)readingData.trend.get(0).sensorTime;
-                m_libreTrendLatest.glucoseLevelRaw = readingData.trend.get(0).glucoseLevelRaw;
+        ListIterator<LibreBlock> li = latestBlocks.listIterator(latestBlocks.size());
+        long lastBlockTime = 0;
+        boolean isLast = true;
+        while (li.hasPrevious()) {
+            LibreBlock libreBlock = li.previous();
+            // Get of the loop if no valid data in the last 15 minutes.
+            if(lastBlockTime == 0) {
+                lastBlockTime = libreBlock.timestamp;
             } else {
-                Log.e(TAG, "Error no readingData.trend for this point, returning withoug doing anything");
-                return m_points;
+                if (JoH.msSince(lastBlockTime, libreBlock.timestamp) > 16 * 60 * 1000) {
+                    // We have readings for the last 16 minutes, but none of them has a BG value.
+                    // This should not happen, but if it does, than it is too much time.
```

Comment:
but if it does, it happens for too long time.

Reply:
Changed.